apiserver/facades/agent/uniter/uniter.go

```diff
@@ -392,8 +410,53 @@ func NewUniterAPIV4(context facade.Context) (*UniterAPIV4, error) {
 	}, nil
 }
 
+// OpenedMachinePortRanges is not available in V16 of the API.
+func (u *UniterAPIV16) OpenedMachinePortRanges(_ struct{}) {}
+
+// OpenedMachinePortRanges returns the port ranges opened by each unit on the
+// provided machines grouped by application endpoint.
+func (u *UniterAPI) OpenedMachinePortRanges(args params.Entities) (params.OpenMachinePortRangesResults, error) {
+	result := params.OpenMachinePortRangesResults{
+		Results: make([]params.OpenMachinePortRangesResult, len(args.Entities)),
+	}
+	canAccess, err := u.accessMachine()
+	if err != nil {
+		return params.OpenMachinePortRangesResults{}, err
+	}
+	for i, entity := range args.Entities {
+		machPortRanges, err := u.getOneMachineOpenedPortRanges(canAccess, entity.Tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+
+		result.Results[i].GroupKey = "endpoint"
+
+		for unitName, unitPortRanges := range machPortRanges.ByUnit() {
+			byEp := make(map[string][]params.PortRange)
+			for endpointName, portRanges := range unitPortRanges.ByEndpoint() {
+				for _, pr := range portRanges {
+					byEp[endpointName] = append(byEp[endpointName], params.FromNetworkPortRange(pr))
+				}
+			}
+			result.Results[i].UnitPortRanges = append(result.Results[i].UnitPortRanges, params.OpenUnitPortRanges{
+				UnitTag:         names.NewUnitTag(unitName).String(),
+				PortRangeGroups: byEp,
+			})
+		}
+
+		// Sort result list by endpoint name to yield consistent results.
+		sort.Slice(result.Results[i].UnitPortRanges, func(a, b int) bool {
+			return result.Results[i].UnitPortRanges[a].UnitTag < result.Results[i].UnitPortRanges[b].UnitTag
+		})
+	}
+	return result, nil
+}
+
 // AllMachinePorts returns all opened port ranges for each given
 // machine (on all networks).
+//
+// DEPRECATED: clients should switch to the OpenedMachinePortRanges API call.
```

Comment:
Why not just go all out and make it not available in V17?   The facade version is changing and there is something to move to, add comments in the OpenedMachinePortRanges.

Reply:
That's actually not a bad idea. I will do that. My next PR (see comment above) also removes the equivalent client method (and some others that we forgot to remove when we switched to CommitHookChanges)

Reply:
Actually, I cannot do this yet as client-side bits still use this API. I have created an API client/facade cleanup card to do this once all my other changes land.